### ISSUE TITLE
Dub method for dependencies

### ISSUE THREAD

**Issue by russel** [mesonbuild/meson#4032]
The dependency search for Dub method dependencies always seems to find that the library was built with dmd even when it was built with ldc2.
```
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
ERROR: Dependency unit-threaded found but it was compiled with dmd and we are using ldc
Dependency unit-threaded found: NO 

meson.build:38:0: ERROR:  Dependency "unit-threaded" not found, tried 
```

Update: building everything with dmd does work, building with ldc2 always fails.

**Comment by GoaLitiuM** [comment]
I can confirm this issue. The method currently used for dub dependency lookup does not work as expected. I talked about the same issue briefly here: https://github.com/mesonbuild/meson/pull/3981#issuecomment-412100702

**Comment by russel** [comment]
It is also worth noting that Meson only uses LDC options for a build. I build the dependency using Dub and DMD, and then set up the project to use DMD (DC must be specified otherwise /usr/bin/ldc2 is used)
```
|> DC=/usr/bin/dmd meson_head --prefix=$HOME/Built ~/Repositories/Git/Masters/GFontBrowser_D
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29)
Dependency gtkd-3 found: YES 3.8.3
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
Dependency unit-threaded found: YES 0.7.50
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
```
but then on building:
```
|> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Dependency gtkd-3 found: YES (cached)
Dependency unit-threaded found: YES (cached)
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
[1/8] Compiling D object 'gfontbrowser@exe/source_GFontBrowser.d.o'.
FAILED: gfontbrowser@exe/source_GFontBrowser.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_GFontBrowser.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/GFontBrowser.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[2/8] Compiling D object 'gfontbrowser@exe/source_main.d.o'.
FAILED: gfontbrowser@exe/source_main.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_main.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/main.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[3/8] Compiling D object 'gfontbrowser@exe/source_About.d.o'.
FAILED: gfontbrowser@exe/source_About.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_About.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/About.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[4/8] Compiling D object 'gfontbrowser-test@exe/source_About.d.o'.
FAILED: gfontbrowser-test@exe/source_About.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_About.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/About.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[5/8] Compiling D object 'gfontbrowser-test@exe/source_main.d.o'.
FAILED: gfontbrowser-test@exe/source_main.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_main.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/main.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[6/8] Compiling D object 'gfontbrowser-test@exe/source_GFontBrowser.d.o'.
FAILED: gfontbrowser-test@exe/source_GFontBrowser.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_GFontBrowser.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/GFontBrowser.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
ninja: build stopped: subcommand failed.
```
Clearly this stems from the meson.build file which has:
```
project(
  'GFontBrowser',
  'd',
  version: '0.0.0',
  default_options: ['buildtype=release'],
)
```
but it appears it is LDC release options that are used even when DC is not LDC.

**Comment by russel** [comment]
Worth noting that "dub describe" assumes debug build with dmd. To get a different response from "dub describe" you need the "--build=relaese" and or "--compiler=/usr/bin/ldc2": the describe command simply reflects the compiler and build type from the command line.

**Comment by GoaLitiuM** [comment]
The wrong optimization levels in #3489 were used for DMD, which broke the release builds. I've included a fix for this in #3981.

The usage of `dub describe` should be rethought here, you can't make any guarantees the user has compiled the depedency with correct compiler. I think better option would be for meson to remove the compiler detection portion from `dub describe`, and invoke `dub build` with the detected compiler to ensure no linking errors would occur.

**Comment by GoaLitiuM** [comment]
I separated the DMD optimization arguments issue to #4060

**Comment by russel** [comment]
I retried the "force using dmd throughout" build:
```
|> DC=/usr/bin/dmd meson_head --prefix=$HOME/Built ~/Repositories/Git/Masters/GFontBrowser_D/ 
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29)
Dependency gtkd-3 found: YES 3.8.3
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
Dependency unit-threaded found: YES 0.7.51
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
 
|> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Dependency gtkd-3 found: YES (cached)
Dependency unit-threaded found: YES (cached)
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
[7/8] Linking target gfontbrowser-test.
FAILED: gfontbrowser-test 
/usr/bin/dmd  -ofgfontbrowser-test 'gfontbrowser-test@exe/source_main.d.o' 'gfontbrowser-test@exe/source_About.d.o' 'gfontbrowser-test@exe/source_GFontBrowser.d.o' -main -L-L/usr/lib/x86_64-linux-gnu/ -L-lgtkd-3 -L-ldl /home/users/russel/.dub/packages/unit-threaded-0.7.51/unit-threaded/libunit-threaded.a  
/usr/bin/ld: /home/users/russel/.dub/packages/unit-threaded-0.7.51/unit-threaded/libunit-threaded.a(unit_threaded.should.o): undefined reference to symbol '_d_allocclass'
/usr/bin/ld: //usr/lib/x86_64-linux-gnu/libdruntime-ldc-shared.so.78: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
[8/8] Linking target gfontbrowser.
ninja: build stopped: subcommand failed.
```

**Comment by FFY00** [comment]
I totally missed this issue (the github issue). When I wrote the patch to add dub support, `dub describe` was updating the `compiler` flag with the last compiler used. This new behavior obviously breaka the compiler check. I'll submit a PR to fix it soon.

**Comment by FFY00** [comment]
This issue should be fixed by #4084 :smile:
